lib/async_hooks.js

```diff
@@ -203,6 +205,141 @@ class AsyncResource {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+const idResourcePairing = new Map();
+const locals = [];
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      for (let i = 0; i < locals.length; ++i) {
+        locals[i]._propagate(resource, triggerResource);
+      }
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      for (let i = 0; i < locals.length; ++i) {
+        delete resource[locals[i].kResourceStore];
+      }
+      idResourcePairing.delete(asyncId);
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
+    asyncContextHook.enable();
+    if (!locals.includes(this)) {
+      locals.push(this);
+    }
+    let resource = getCurrentResource();
+    if (!resource) {
+      // There is no resource, we need to create a fake one
+      resource = {};
+      idResourcePairing.set(executionAsyncId(), new WeakReference(resource));
+      process.nextTick(() => {
+        resource;
+      });
+      // This prevents the fake resource from being garbage collected too soon
+    }
+    const store = new Map();
+    resource[this.kResourceStore] = store;
+    return store;
+  }
+
+  _exit() {
+    const resource = getCurrentResource();
+    if (resource) {
+      delete resource[this.kResourceStore];
+    }
+  }
+
+  runSyncAndReturn(callback, ...args) {
+    this._enter();
+    try {
+      return callback(...args);
+    } finally {
+      this._exit();
+    }
+  }
+
+  exitSyncAndReturn(callback, ...args) {
+    const resource = getCurrentResource();
+    if (!resource) {
+      return callback(...args);
+    }
+    const store = resource[this.kResourceStore];
+    try {
+      return callback(...args);
+    } finally {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const resource = getCurrentResource();
+    if (resource) {
+      return resource[this.kResourceStore];
+    }
+  }
+
+  /**
+   * after this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  run(callback, ...args) {
+    this._enter();
+    process.nextTick(() => callback(...args));
```

Comment:
I wonder what's the point in using `process.nextTick` here if scoping (`._enter()`/`._exit()`) is done synchronously?

Reply:
Mostly error management, here there is only 1 path for errors. If anything in the callback throws, it goes to a global handler executed within the context.

with `runSyncAndReturn` there are two paths:
* an error thrown in the the callback goes outside of the `runSyncAndReturn` where the store is not available
* an error thrown within an async operation created from the callback will behvae as in the `run` case.

Reply:
> Mostly error management, here there is only 1 path for errors. If anything in the callback throws, it goes to a global handler executed within the context.

So, the only difference is error handling. That feature can be easily implemented by users who want this specific behavior (and I think there won't be a lot of them) by wrapping `.run()` into, say, a `process.nextTick()`. With the current implementation all users will be "paying" that additional `process.nextTick`.

Reply:
That's actually a big chunk IMO as I plan to use this API as a basis to help users moving outside of domains to a safer error management system :/

Reply:
I wouldn't call it a safer error management system. It may be handy for APMs, but not for all users.

If you believe that this is the right approach, then why there is a sync `.runSyncAndReturn()` method, not just `.runAndReturn()`?

Reply:
You mean a `runAndReturn` that runs something async and return its result? 🤔 I think you lost me here 😕

Reply:
I mean the following. Currently the API exposes two methods (`.run()` and `.runSyncAndReturn`). If you think that `.run()` is way better in terms of error management, why there is `.runSyncAndReturn()` method, which is synchronous, instead of, say, `.runAndReturn()`?

Reply:
`runAndReturn` can't really work as it goes in an asynchronous flow (except if we are going with promises only). `runSyncAndReturn` is here to cover the cases when `run` can't be used - by instrumentation providers for instance.

Reply:
Yes, my question was about promise (or additional callback which ugly) based `.runAndReturn()` method. Anyway, thanks for the clarification on the rationale behind async `.run()` method. That was my question.

Reply:
Awesome! Sorry I had a hard time getting it 😄